chainer/testing/backend.py

```diff
@@ -0,0 +1,126 @@
+import functools
+import unittest
+
+import numpy
+
+import chainer
+from chainer import cuda
+from chainer.testing import attr
+
+
+class BackendConfig(object):
+
+    _props = [
+        ('use_cuda', False),
+        ('use_cudnn', 'never'),
+        ('cudnn_deterministic', False),
+        ('autotune', False),
+    ]
+
+    def __init__(self, params):
+        assert isinstance(params, dict)
+        self._contexts = []
+
+        # Default values
+        for k, v in self._props:
+            setattr(self, k, v)
+        # Specified values
+        for k, v in params.items():
+            if not hasattr(self, k):
+                raise ValueError('Parameter {} is not defined'.format(k))
+            setattr(self, k, v)
+
+    @property
+    def xp(self):
+        if self.use_cuda:
+            return cuda.cupy
+        else:
+            return numpy
+
+    def __enter__(self):
+        self._contexts = [
+            chainer.using_config(
+                'use_cudnn', self.use_cudnn),
+            chainer.using_config(
+                'cudnn_deterministic', self.cudnn_deterministic),
+            chainer.using_config(
+                'autotune', self.autotune),
+        ]
+        for c in self._contexts:
+            c.__enter__()
+        return self
+
+    def __exit__(self, typ, value, traceback):
+        for c in reversed(self._contexts):
+            c.__exit__(typ, value, traceback)
+
+    def __repr__(self):
+        lst = []
+        for k, _ in self._props:
+            lst.append('{}={!r}'.format(k, getattr(self, k)))
+        return '<BackendConfig {}>'.format(' '.join(lst))
+
+    def get_func_str(self):
+        """Returns a string that can be used in method name"""
+        lst = []
+        for k, _ in self._props:
+            val = getattr(self, k)
+            if val is True:
+                val = 'true'
+            elif val is False:
+                val = 'false'
+            else:
+                val = str(val)
+            lst.append('{}_{}'.format(k, val))
+        return '__'.join(lst)
+
+    def get_pytest_marks(self):
+        marks = []
+        if self.use_cuda:
+            marks.append(attr.gpu)
+            if self.use_cudnn != 'never':
+                marks.append(attr.cudnn)
+
+        assert all(callable(_) for _ in marks)
```

Comment:
Do we need this assertion?

Reply:
Yes, to detect a glitch in the preceding logic earlier.

Reply:
Okay, then an assertion is appropriate.